lib/ts/recipe/accountlinking/recipe.ts

```diff
@@ -799,4 +844,397 @@ export default class Recipe extends RecipeModule {
             }
         }
     };
+
+    private async shouldBecomePrimaryUser(
+        user: User,
+        tenantId: string,
+        session: SessionContainerInterface | undefined,
+        userContext: UserContext
+    ) {
+        const shouldDoAccountLinking = await this.config.shouldDoAutomaticAccountLinking(
+            user.loginMethods[0],
+            undefined,
+            session,
+            tenantId,
+            userContext
+        );
+
+        if (!shouldDoAccountLinking.shouldAutomaticallyLink) {
+            logDebugMessage("shouldBecomePrimaryUser returning false because shouldAutomaticallyLink is false");
+            return false;
+        }
+
+        if (shouldDoAccountLinking.shouldRequireVerification && !user.loginMethods[0].verified) {
+            logDebugMessage(
+                "shouldBecomePrimaryUser returning false because shouldRequireVerification is true but the login method is not verified"
+            );
+            return false;
+        }
+
+        logDebugMessage("shouldBecomePrimaryUser returning true");
+
+        return true;
+    }
+
+    private async tryLinkingBySession({
+        sessionUserLinkingRequiresVerification,
+        authLoginMethod,
+        inputUser,
+        sessionUser,
+        tenantId,
+        userContext,
+    }: {
+        inputUser: User;
+        sessionUserLinkingRequiresVerification: boolean;
+        sessionUser: User;
+        authLoginMethod: LoginMethod;
+        tenantId: string;
+        userContext: UserContext;
+    }): Promise<
+        | { status: "OK"; user: User }
+        | { status: "LINKING_TO_SESSION_USER_FAILED" }
+        | { status: "NON_PRIMARY_SESSION_USER" }
+    > {
+        logDebugMessage("tryLinkingBySession called");
+        const usersToLink = await this.getUsersThatCanBeLinkedToRecipeUser({
+            tenantId,
+            user: inputUser,
+            userContext,
+        });
+        const primaryUserThatCanBeLinkedToTheInputUser = usersToLink.primaryUser;
+
+        // If we get here, we know that the session user is primary.
+        // if both of them are primary, they cannot be linked together.
+        // we also check this outside of this function.
+        if (inputUser.isPrimaryUser) {
+            return { status: "LINKING_TO_SESSION_USER_FAILED" };
+        }
+
+        // If the input user has another user (an it's not the session user) it could be linked to based on account info
+        // then we can't link it to the session user.
+        if (
+            primaryUserThatCanBeLinkedToTheInputUser !== undefined &&
+            sessionUser.id !== primaryUserThatCanBeLinkedToTheInputUser.id
+        ) {
+            return { status: "LINKING_TO_SESSION_USER_FAILED" };
+        }
```

Comment:
why do we need to do this check? When we call linkAccounts, it will fail anyway.